text/0000-contract-metadata.md

```diff
@@ -0,0 +1,117 @@
+- Proposal Name: `contract_metadata`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no direct way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the methods name and parameters are already mangled.
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
+especially in the case of financial apps, they have no choice but to hardcode them in the front end, which is suboptimal in many ways.
+Furthermore, if developers want to get the list of methods in some contract for some downstream task like data analysis,
+they have no way of doing so. Contract metadata aims to solve the aforementioned problems by providing a convenient way 
+for developers to list the methods of contracts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For developers, there will be two main changes:
+- Instead of annotating view and change methods in the `initContract` function in `main.js`,
+they will instead annotate the methods of a contract by decorators.
+More specifically, every method is by default a change method, unless annotated by `@view_method`.
```

Comment:
I suggest that all our methods are callable as `near call` and `near view` independently on whether they change the state or not. We should remove the current constraint the prohibits calling `near view` on methods that mutate the state. There are not safety concerns to have this constraint. @evgenykuzyakov @ilblackdragon

Reply:
it's not about safety, it's about developer experience. If I call the function and it didn't purist state change silently I won't know WTF is going on.

Reply:
If you are calling `near view` then you know that it has effect no permanent effect on the state, because you read  `near view --help` before calling it :)

Reply:
On this point, does a `view` contract function have the restriction that it can't call a `non-view` function?

Reply:
Yes. Right now if you call a state-changing function within a view function, the execution will fail.

Reply:
So calling a state-changing function with `near view` is fine, but calling a `view` function that calls a state changing function (with `near view`) is not?

Reply:
No. `near view` is the proposed change, not the status quo.

Reply:
I'm saying that right now, it is an error to call a `view` function in a `non-view` function, so should it not be the same for `near view` to throw an error when calling a `non-view` function?

Reply:
I think you misunderstood. Calling a view function in a non-view function is fine, the converse is not.

Reply:
That was a typo yes, 

Calling a `non-view` function with a `view` one, should error.
Therefore, calling a `non-view` function with `near view` should be an error as well?

Reply:
In the proposed change by max, that will not error but the state will not be changed instead.

Reply:
The two should be consistent though? Either they both error or neither error?

Reply:
I still think there is some advantage of being explicit. The annotation serves as documentation for developers who might interact with the contract and with metadata, the list of view methods can be easily extracted and displayed, which is better, in my opinion, than looking at the source code and tries to understand for each function, whether it changes the state or not.

Reply:
From today's hackathon, the distinction between view and change methods should be either strictly enforced (and extensively documented), or their functionality should be merged into a single kind. Otherwise, it gets really confusing and error-prone.

Reply:
Okay it seems like the distinction is confusing. I think we should eliminate the difference and allow dry run of change methods. One thing that is not clear to me is how we should support cross contract view calls. It seems that if you pass 0 gas, the instantiation of wasm module for the call on the second contract will fail.